From ffc10f12412553e61ab2c36b295ee61b033e11df Mon Sep 17 00:00:00 2001
From: "William Beasley (The Capable Hub)" <wbeasley@thegoodpenguin.co.uk>
Date: Tue, 14 Apr 2026 23:14:00 +0100
Subject: [PATCH 3/5] cheri: Increase size of argument buffer and align

CHERI pointers are capabilities and are larger than other platforms.
They also require 16-byte alignment.

Signed-off-by: William Beasley (The Capable Hub) <wbeasley@thegoodpenguin.co.uk>
---
 wolfssl/internal.h | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/wolfssl/internal.h b/wolfssl/internal.h
index 7ef18d6c7..6504f467d 100644
--- a/wolfssl/internal.h
+++ b/wolfssl/internal.h
@@ -5761,7 +5761,7 @@ typedef struct BuildMsgArgs {
 #endif
 
 #ifdef WOLFSSL_ASYNC_IO
-    #define MAX_ASYNC_ARGS 18
+    #define MAX_ASYNC_ARGS 24
     typedef void (*FreeArgsCb)(struct WOLFSSL* ssl, void* pArgs);
 
     struct WOLFSSL_ASYNC {
@@ -5769,7 +5769,11 @@ typedef struct BuildMsgArgs {
         BuildMsgArgs  buildArgs; /* holder for current BuildMessage args */
 #endif
         FreeArgsCb    freeArgs; /* function pointer to cleanup args */
+#ifdef __CHERI_PURE_CAPABILITY__
+        max_align_t args[MAX_ASYNC_ARGS * sizeof(word32) / sizeof(max_align_t)]; /* holder for current args */
+#else
         word32        args[MAX_ASYNC_ARGS]; /* holder for current args */
+#endif
     };
 #endif
 
-- 
2.43.0

